User table refactor 1

javitonino opened the pull request:
See #11505 

**Risk meter:** 🔴🔴⚪️⚪️⚪️
Old model is untouched except for moving some methods, validations added to new model should have little effect, since it is not used for updates

^ Introducing risk factor! my personal way to try to visually indicate how risky a PR is (how much testing it needs) and providing a quick explanation.

---

My approach here is to change `::Table` to use the AR model, instead of the Sequel one. Then run `spec/models/table_spec.rb` and see what breaks.

Problems encountered:
- `Sequel.new?` vs `ActiveRecord.new_record?`. I added a `def new? new_record? end` in the AR model to keep working. Removed in the final version.
- Table controls map canonical map creation, so we need to move all that to the new models at once (transaction issues). This is going to be fun (tight coupling of Table and UserTable, specially, model hooks)

To address the second problem, I started moving stuff from the table service to the models when it makes sense, in order to make more explicit the separation that the service works on the physical table, and the model on the metadata.

In this PR:
- Moved validation to both models.
- Moved canonical visualization creation to the old model. Not yet to the new one.
- Spec fixes to account for the new AR model validation. Note the factory change (setting privacy to nil). That is because AR default to 0 (from DB) instead of nil (like Sequel), so `set_default_privacy` has no effect (as privacy is preset to 0). We should look into what is a good solution for this, depending on where this is used.
- Tests for user table validation in the model specs (there is some duplication with tests in `table_spec`)

Next steps:
- Implement `after_create` in the AR model, which basically consists of creating the canonical visualization. This is where most of the tests are failing now (when replacing the model in `::Table.initialize`).

---

**Acceptance**
It's hard to specify an acceptance for this. Here are some suggestions that cover key operations around the refactored code:
- [x] Import a dataset
- [x] Import a new empty dataset
- [x] Open the previous datasets in the public and private views.
- [x] Create a table via SQL API and let ghost tables detect it
- [x] Change privacy of a table
- [x] Rename a table
- [x] Create a map, then delete the related dataset (the map should be deleted)
- [x] Share a table with another user. Open that table with the other user account.
- [x] Import a dataset in "Maps" section and open the visualization, both in Builder and public page.
- [x] Sync a dataset. Force update afterwards.

javitonino commented on app/models/table/user_table.rb:
```diff
@@ -295,4 +315,60 @@ def estimated_row_count
+  def create_default_map_and_layers
+    base_layer = ::ModelFactories::LayerFactory.get_default_base_layer(user)
+
+    self.map = ::ModelFactories::MapFactory.get_map(base_layer, user_id, id)
```
This method is a direct copy from the original, except for this line. It changes `map_id` to `map`

juanignaciosl replied:
Is using `user_id` needed? Using both `user.id` and `user_id` is usually a source of troubles, because they look the same but they're not. IMHO, in 1-1 relationships, the relation should be used (`user.id`) unless there's an explicit need. If it's intended, I'd document why.

javitonino replied:
Just moving code, so I did not change anything. I will try to change it, as I agree it's usually better to use the model.

javitonino commented on app/models/table/user_table.rb:
```diff
@@ -295,4 +315,60 @@ def estimated_row_count
+
+    member = CartoDB::Visualization::Member.new(
+      name:         name,
+      map_id:       map.id,
```
This method is a direct copy from the original, except for this line. It changes `map_id` to `map.id`

javitonino commented on spec/factories/user_tables.rb:
```diff
@@ -9,6 +9,7 @@
 
   factory :carto_user_table, class: Carto::UserTable do
     name { unique_name('user_table') }
+    privacy nil
```
Needed because AR takes the default from DB, and sets `privacy = 0`, which breaks `set_default_privacy` on `before_validation`.

juanignaciosl replied:
That might better be a code comment instead of a GitHub one. There's something weird there, though. How come don't we want to mimic AR normal behaviour? This would fix the behaviour for the tests but there might be troubles with production code `Carto::UserTable` new instances, isn't it?

javitonino replied:
Agreed with moving the comment to the source.

About the behaviour. When creating a new table without specifying privacy, it should be set to `default_table_privacy`, which calculates it depending on user type, checking if private tables are enabled or not. This is implemented in the model in a `before_validation` hook, that sets the privacy to the appropriate value if it is nil.

Now, the problem, AR insists on writing the DB default (`0` for private table), so the hook is not triggered. So if the user doesn't have private tables enabled, the validation will fail ("unauthorized to create private tables"). Now, my desired solution would be to override the DB default in the model, but [the method to do that](http://edgeapi.rubyonrails.org/classes/ActiveRecord/Attributes/ClassMethods.html#method-i-attribute) is not available until Rails 4.

So I went with this, but I don't like it very much, which is why I left the bait comment, to discuss it. I don't like it because we would have to specify `privacy: nil` in pretty much all the user table creations (e.g: from the importer). I tried to avoid it with some initialization hook, but I find it impossible to differentiate the case where the `0` comes from the caller (`Carto::UserTable.new(privacy: PRIVACY_PRIVATE)`) against the case where this is the default (`Carto::UserTable.new()`). 

The "killing flies with cannons" way is to remove the DB default, but it definitely seems like overkill.

Extra comment: this attribute is redundant, duplicated with canonical visualization privacy. There is code to sync them, since different places of the code check in the different models.

juanignaciosl replied:
What puzzles me is not the test but the actual code. You say:

> AR insists on writing the DB default (0 for private table), so the hook is not triggered. 

Then, how come the hook works for production code? I could understand a `privacy 0` here to make sure that tests test the AR drawback. Nevertheless, `privacy nil` seems to fix it (but only for tests).

javitonino replied:
Obviously it doesn't work for production code neither. Right now, we are not creating any `Carto::UserTable` from production code, so it's not an issue. That's why I say "I don't like it because we would have to specify privacy: nil in pretty much all the user table creations (e.g: from the importer).", where I was referring to production code (sorry, rereading, it's not clear).

So yeah, this sucks because we would have to do this same thing in production code all over the place.

juanignaciosl commented on app/models/carto/user_table.rb:
```diff
@@ -25,6 +26,15 @@ class UserTable < ActiveRecord::Base
     has_many :layers_user_table
     has_many :layers, through: :layers_user_table
 
+    validates :user_id, presence: true
```
I think that this is wrong for creations. Assigning a `user` instead of a `user_id` might make this fail, although it should be valid.

javitonino replied:
It works as expected (you don't have to trust me, there is a test for that!).
```
ut=Carto::UserTable.new(user: Carto::User.first)
ut.valid?
=> true
```

It seems the validation it's the same, independent of what you specify (`user_id` or `user`). That said, it's probably better to write `user`. :+1:

juanignaciosl commented on app/models/carto/user_table.rb:
```diff
@@ -25,6 +26,15 @@ class UserTable < ActiveRecord::Base
+    validates :privacy, inclusion: [PRIVACY_PRIVATE, PRIVACY_PUBLIC, PRIVACY_LINK].freeze
+    validate :validate_privacy_changes
+
+    before_validation :set_default_table_privacy
```
Minor/just a pun: `before_validation` might be more clear _before validations_ :)

juanignaciosl commented on spec/models/carto/user_table_spec.rb:
```diff
@@ -8,7 +8,7 @@
     @user = FactoryGirl.create(:carto_user)
     @carto_user = @user
-    @user_table = Carto::UserTable.create(user: @user, name: 'user_table')
+    @user_table = FactoryGirl.create(:carto_user_table, user: @user)
```
I'm not sure whether `user_table_spec` should use a FactoryGirl `Carto::UserTable` instance. In a test, the SUT should normally be instantiated "the normal way".

juanignaciosl commented on spec/models/carto/visualization_spec.rb:
```diff
@@ -104,6 +104,7 @@
     include Carto::Factories::Visualizations
 
     it 'only returns tables that a user can read' do
+      @carto_user.update_attribute(:private_tables_enabled, true)
```
How come this line is needed here? Nothing else changes within this test, it's strange.

javitonino replied:
The old test was bugged, but it did pass because no validations were made in `Carto::UserTable`. I expect this to happen more and more as we add functionality, as we got in the habit of using `Carto::` models when we want to skip validations. The problem was the line:
`private_table = FactoryGirl.create(:private_user_table, user: @carto_user)`
which should obviously fail if the user doesn't have private tables enabled.

javitonino commented:
Addressed the CR concerns. I found a different way for the `privacy: nil` default, by overriding the [`column_defaults`](http://apidock.com/rails/v3.2.13/ActiveRecord/ModelSchema/ClassMethods/column_defaults) method in order to use the default we want.

It's doing something similar to what [`attribute`](http://api.rubyonrails.org/v5.0.0/classes/ActiveRecord/Attributes/ClassMethods.html) does (Rails 5+), but as we are touching an AR method, I added an explicit test.

juanignaciosl commented on app/models/carto/user_table.rb:
```diff
@@ -13,6 +14,10 @@ class UserTable < ActiveRecord::Base
     }.freeze
 
+    def self.column_defaults
+      super.merge("privacy" => nil)
```
Minor: much better fix, this will not only fix the test but also production code 👍 

I think that the comment at fc47437f7c035ac46aa54b7ef0b91f7560810c96 is still valuable, though.

juanignaciosl commented on spec/models/carto/user_table_spec.rb:
```diff
@@ -19,7 +21,15 @@
+    end
+  end
+
+  it 'sets privacy to nil by default' do
```
Minor: in addition, and to make more clear the need for the default to nil, it'd be nice checking that an explicit `privacy` set at construction works, and that passing a user defaults to the right private/public value. Mixed feelings, because it'd be duplicated in shared examples. Feel free to ignore.

javitonino replied:
Uhm, it seems that there are no tests for default privacy based on user settings. I'll add some.

juanignaciosl commented:
Acceptance suggestions (WIP, move to the first comment when this goes to acceptance):

- [ ] After importing the dataset, open its public page.

juanignaciosl commented on app/models/carto/user_table.rb:
```diff
@@ -138,6 +160,28 @@ def table
-  end
 
+    def validate_user_not_viewer
+      errors.add(:user, "Viewer users can't create tables") if user && user.viewer
```
Minor: you can use `user.try(:viewer)` instead.

juanignaciosl commented on app/models/carto/user_table.rb:
```diff
@@ -138,6 +160,28 @@ def table
+        end
+
+        # if the table exists, is private, but the owner no longer has private privileges
+        if !new_record? && private? && privacy_changed?
```
Minor: can't both checks be merged into `!new_record && !public?`?

javitonino replied:
I think we can do even better. Refactoring...

juanignaciosl commented on app/models/table.rb:
```diff
@@ -519,49 +480,6 @@ def default_baselayer_for_user(user=nil)
     end
   end
 
-  def create_default_map_and_layers
```
I think that deleting this breaks the call from `migrator20.rb`.

javitonino replied:
Yup, how do you feel about removing those migrators altogether? They are very old migrations and depend on the models and they are very broken. For example, the migration you pointed to does `::Table.all`, which doesn't work in current code (as `::Table` is no longer a model).

juanignaciosl replied:
🔥

juanignaciosl commented on app/models/table/user_table.rb:
```diff
@@ -159,28 +159,43 @@ def validate
-    errors.add(
-      :name, 'is a reserved keyword, please choose a different one'
-    ) if Carto::DB::Sanitize::RESERVED_TABLE_NAMES.include?(name)
+    if Carto::DB::Sanitize::RESERVED_TABLE_NAMES.include?(name)
```
:facepalm:

juanignaciosl commented:
@javitonino ey, migrations can't be deleted, migrate process won't run. I overlooked  `db/migrate/20121017152007_migrate_data_from_1_0_to_2_0.rb` deletion on CR because I omitted most deletions. Others seem right.

javitonino commented:
Oh, that's why it was commented. :facepalm: 

Thanks ❤️
